### PR TITLE
#mission_nps_mobile_40: Refactor `capture-receipt` component

### PR DESCRIPTION
**Pull request by yksolanki9**
Design doc - https://www.notion.so/fyleuniverse/Capture-Receipt-Component-Refactoring-8acad141897c4e3881353ab387dc0c87

- In this PR, some logic from `capture-receipt` component is refactored and moved into the new `camera-preview` component
- The changes are tested on android and ios
- Will do some refactoring in `capture-receipt` component in the next PR.

### CODE REVIEW

File: src/app/shared/components/capture-receipt/capture-receipt.component.ts [hunk 1]
```diff
@@ -1,8 +1,7 @@
-import { Component, EventEmitter, OnDestroy, OnInit, Input, AfterViewInit } from '@angular/core';
-import { CameraPreview, CameraPreviewOptions, CameraPreviewPictureOptions } from '@capacitor-community/camera-preview';
-import { Capacitor } from '@capacitor/core';
+import { Component, EventEmitter, OnDestroy, OnInit, Input, AfterViewInit, ViewChild } from '@angular/core';
```
Comment (yksolanki9) [inline src/app/shared/components/capture-receipt/capture-receipt.component.ts:1]: This file will be cleaned in the next PR, so ignore issues in the unchanged code

**Review by Julias0 — CHANGES_REQUESTED** [review]
Added a few comments

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        this.loaderService.hideLoader();
+      });
+    }
+  }
+
+  async stopCamera() {
+    if (this.isCameraPreviewInitiated) {
+      this.isCameraPreviewInitiated = false;
+      await CameraPreview.stop();
+      this.isCameraPreviewStarted = false;
```
Comment (Dimple16) [inline src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts:104]: can assign this value after the `.then` of `CameraPreview.stop()`?

**Review by Dimple16 — CHANGES_REQUESTED** [review]

File: src/app/shared/components/capture-receipt/capture-receipt.component.ts [hunk 3]
```diff
@@ -30,6 +29,8 @@ type Image = Partial<{
   styleUrls: ['./capture-receipt.component.scss'],
 })
 export class CaptureReceiptComponent implements OnInit, OnDestroy, AfterViewInit {
+  @ViewChild('cameraPreview') cameraPreview: CameraPreviewComponent;
+
   @Input() isModal = false;
 
   @Input() allowGalleryUploads = true;
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/capture-receipt.component.ts:36]: Yes, in android when adding receipts from inside expense forms

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.html [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    <div *ngIf="allowBulkFyle" class="camera-preview__bottom-container__icon-block text-center">
+      <ng-container *ngIf="noOfReceipts === 0; else receiptPreviewIcon">
+        <div (click)="onSwitchMode()">
+          <img
+            class="camera-preview__bottom-container__icon-block__icon"
+            [src]="isBulkMode ? '../../../../assets/images/single-mode.png' : '../../../../assets/images/bulk-mode.png'"
+            alt="Mode"
+          />
+          <div class="camera-preview__bottom-container__text--secondary">
+            {{ !isBulkMode ? 'Bulk Mode' : 'Single Mode' }}
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.html:61]: `!isBulkMode ? 'Bulk Mode' : 'Single Mode'` is correct
If user is not on bulk mode, show `Bulk mode` on the bottom right icon to prompt him/her to switch to bulk mode

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+  @Output() toggleFlashMode = new EventEmitter();
+
+  @Output() dismissCameraPreview = new EventEmitter();
+
+  isCameraPreviewInitiated = false;
+
+  isCameraPreviewStarted = false;
+
+  flashMode: string;
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts:46]: Done

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+  hasModeChanged = false;
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts:48]: Done

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    from(Camera.requestPermissions()).subscribe((permissions) => {
+      if (permissions?.camera === 'denied') {
+        return this.showPermissionDeniedPopover('CAMERA');
+      }
+
+      /*
+       * 'prompt-with-rationale' means that the user has denied permission, but has not disabled the permission prompt.
+       * So, we can use the native dialog to ask the user for camera permission.
+       */
+      if (permissions?.camera === 'prompt-with-rationale') {
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts:71]: There's a return statement above so it doesn't matter

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+      this.isCameraPreviewInitiated = false;
+      await CameraPreview.stop();
+      this.isCameraPreviewStarted = false;
+    }
+  }
+
+  getFlashModes() {
+    if (Capacitor.getPlatform() !== 'web') {
+      CameraPreview.getSupportedFlashModes().then((flashModes) => {
+        if (flashModes.result && flashModes.result.includes('on') && flashModes.result.includes('off')) {
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts:111]: Done

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts (hunk 2 above)
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts:104]: The idea is to get rid of promises altogether, converting it into an observable

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+      this.loaderService.showLoader();
+      CameraPreview.start(cameraPreviewOptions).then((res) => {
+        this.isCameraPreviewStarted = true;
+        this.getFlashModes();
+        this.loaderService.hideLoader();
+      });
+    }
+  }
+
+  async stopCamera() {
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts:100]: Done

File: src/app/shared/components/capture-receipt/capture-receipt.component.ts [hunk 10]
```diff
@@ -532,9 +404,9 @@ export class CaptureReceiptComponent implements OnInit, OnDestroy, AfterViewInit
       } else {
         from(Camera.requestPermissions({ permissions: ['photos'] })).subscribe((permissions) => {
           if (permissions?.photos === 'denied') {
-            return this.showPermissionDeniedPopover('GALLERY');
+            return this.cameraPreview.showPermissionDeniedPopover('GALLERY');
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/capture-receipt.component.ts:407]: Yes, done

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+  constructor(
+    private loaderService: LoaderService,
+    private popoverController: PopoverController,
+    private platform: Platform
+  ) {}
+
+  setUpAndStartCamera() {
+    if (Capacitor.getPlatform() === 'web') {
+      this.startCameraPreview();
+      return;
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts]: Replaced with `else if`

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+
+  @Output() receiptPreview = new EventEmitter();
+
+  @Output() toggleFlashMode = new EventEmitter();
+
+  @Output() dismissCameraPreview = new EventEmitter();
+
+  isCameraPreviewInitiated = false;
+
+  isCameraPreviewStarted = false;
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts]: Have replaced them with an enum

File: src/app/constants.ts [hunk 13]
```diff
@@ -1,3 +1,5 @@
 import { InjectionToken } from '@angular/core';
 
 export const PAGINATION_SIZE = new InjectionToken<number>('API Pagination size');
+
+export const DEVICE_PLATFORM = new InjectionToken<'android' | 'ios' | 'web'>('Device Platform');
```
Comment (yksolanki9) [inline src/app/constants.ts:5]: Will create an enum for this and use this injection across the app in a separate pr as it is out of scope for this change

File: src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+  hasModeChanged = false;
+
+  constructor(
+    private loaderService: LoaderService,
+    private popoverController: PopoverController,
+    private platform: Platform
+  ) {}
+
+  setUpAndStartCamera() {
+    if (Capacitor.getPlatform() === 'web') {
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/capture-receipt/camera-preview/camera-preview.component.ts:60]: Used dependency injection as discussed

**Review by Dimple16 — APPROVED** [review]

**Review by Julias0 — APPROVED** [review]

**Pull request merged**